Password manager UI can cause conditional requests to pause

pascoej opened the pull request:
#### d691e4579511b2fe0f916d9da1746f98b0519036
<pre>
Password manager UI can cause conditional requests to pause
<a href="https://bugs.webkit.org/show_bug.cgi?id=270135">https://bugs.webkit.org/show_bug.cgi?id=270135</a>
<a href="https://rdar.apple.com/123659147">rdar://123659147</a>

Reviewed by Charlie Wolfe and Chris Dumez.

The API for autofill assisted passkey requests only allows one request ongoing per UI process.
In order to handle this behavior, we currently pause requests whenever their page is not active
and resume them whenever the page becomes active.

However, bringing up password manager UI as part of an conditional mediation may cause the page to
become not active, therefore pausing the request the user is selecting a credential for in the UI.

To fix this issue, we don&apos;t pause requests whenever a page becomes inactive. Instead we pause the active
request whenever another page becomes active. This will result in no pause when brining up password manager
UI.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::pauseConditionalAssertion):
(WebKit::WebAuthenticatorCoordinatorProxy::makeActiveConditionalAssertion):
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchActivityStateChange):

Canonical link: <a href="https://commits.webkit.org/275414@main">https://commits.webkit.org/275414@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/840d79d0a436180ca7e87c72ee436de4ddd715e1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41783 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20797 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44165 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44360 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/37876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23940 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18128 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42357 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17725 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35977 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/15201 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45754 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/37969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37323 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41079 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/16598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/39523 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/18217 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9365 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/18275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -360,6 +384,8 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
         return;
     }
+    if (m_isConditionalMediation)
+        activeConditionalMediationProxy() = WeakPtr { *this };
```
nit: `WeakPtr { }` should be unnecessary here.

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -320,13 +320,24 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
+WeakPtr<WebAuthenticatorCoordinatorProxy>& WebAuthenticatorCoordinatorProxy::activeConditionalMediationProxy()
 {
-    if (m_paused)
+    static NeverDestroyed<WeakPtr<WebAuthenticatorCoordinatorProxy>> proxy;
```
Can we use `MainThreadNeverDestroyed`?

charliewolfe commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -339,6 +350,19 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
 
+void WebAuthenticatorCoordinatorProxy::makeActiveConditionalAssertion()
+{
+    if (auto activeProxy = activeConditionalMediationProxy()) {
```
```suggestion
    if (auto& activeProxy = activeConditionalMediationProxy()) {
```

charliewolfe commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -320,13 +320,24 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
+{
+    if (m_paused) {
+        completionHandler();
         return;
```
Could be `return completionHandler();`

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -320,13 +320,24 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
     m_paused = true;
-    if (m_isConditionalMediation)
+    if (m_isConditionalMediation) {
+        m_cancelHandler = WTFMove(completionHandler);
```
What if something is waiting for `m_cancelHandler` to be called? Should we trigger it before reassigning it?

brentfulgham replied:
It looks like in `WebAuthenticatorCoordinatorProxy::cancel` when we pass a new handler we just overwrite any cancel handler already in existence, so perhaps no special handling is needed here.

pascoej committed:
Password manager UI can cause conditional requests to pause
https://bugs.webkit.org/show_bug.cgi?id=270135
rdar://123659147

Reviewed by Charlie Wolfe and Chris Dumez.

The API for autofill assisted passkey requests only allows one request ongoing per UI process.
In order to handle this behavior, we currently pause requests whenever their page is not active
and resume them whenever the page becomes active.

However, bringing up password manager UI as part of an conditional mediation may cause the page to
become not active, therefore pausing the request the user is selecting a credential for in the UI.

To fix this issue, we don't pause requests whenever a page becomes inactive. Instead we pause the active
request whenever another page becomes active. This will result in no pause when brining up password manager
UI.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::pauseConditionalAssertion):
(WebKit::WebAuthenticatorCoordinatorProxy::makeActiveConditionalAssertion):
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchActivityStateChange):

Canonical link: https://commits.webkit.org/275414@main
